### PR TITLE
Routing region fix #7668

### LINKED ISSUE #7668: Roundabout preview wrong in Mauritius for left-handed traffic
Hey,

the following occured on a left-handed traffic country (Mauritius):

Screenshot_20191012-185043 943

The routing itself is correct, but the preview of the roundabout (upper left) is generated as it would be a right-handed country.

Route through roundabout in Mauritius (screenshot):
https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=-20.02182%2C57.58291%3B-20.02060%2C57.58143

Using the following OSMAnd version:
**OsmAnd+ 3.4.8** on Android (Nokia 8 Sirocco)

Using the following offline maps:
- Germany Hamburg 01.10.2019
- Mauritius 01.10.2019
- World overview 16.07.2019
- World basemap mini 22.08.2019

The Osmand setting "traffic region" was set to automatic.

Everything else was working fine except the roundabout direction preview 😃

### PR DESCRIPTION
**Pull request by MadWasp79**

### CODE REVIEW

File: OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java [hunk 1]
```diff
@@ -57,6 +57,7 @@
 	private List<LatLon> intermediatePoints;
 	private Location lastProjection;
 	private Location lastFixedLocation;
+	private LatLon lastStartLocation = null;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: should be stored offfline in settings

File: OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java [hunk 2]
```diff
@@ -260,6 +262,17 @@ public LatLon getFinalLocation() {
 		return finalLocation;
 	}
 
+	public void cacheStartLocation(Location nextStartLocation) {
+		LatLon start = new LatLon(nextStartLocation.getLatitude(), nextStartLocation.getLongitude());
+		if (lastStartLocation == null) {
+			lastStartLocation = new LatLon(nextStartLocation.getLatitude(), nextStartLocation.getLongitude());
+			app.getMapViewTrackingUtilities().detectDrivingRegion(lastStartLocation);
+		} else if (MapUtils.getDistance(start, lastStartLocation) > 100000) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: Constant ?

Comment (vshcherb) [reply OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: Could be 1 line "lastStartLocation == null || MapUtils.getDistance(start, lastStartLocation) < CACHE_RADIUS )

File: OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java [hunk 3]
```diff
@@ -260,6 +262,17 @@ public LatLon getFinalLocation() {
 		return finalLocation;
 	}
 
+	public void cacheStartLocation(Location nextStartLocation) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: cacheStartLocation method name is wrong ?

File: OsmAnd-java/src/main/java/net/osmand/map/OsmandRegions.java [hunk 4]
```diff
@@ -743,17 +743,8 @@ public BinaryMapDataObject getSmallestBinaryMapDataObjectAt(List<BinaryMapDataOb
...
-						TagValuePair tp = o.getMapIndex().decodeType(o.getTypes()[i]);
-						if ("boundary".equals(tp.value)) {
-							isRegion = false;
-							break;
-						}
-					}
 					WorldRegion downloadRegion = getRegionData(getFullName(o));
-					if (!isRegion
-							|| downloadRegion == null
+					if ( downloadRegion == null
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/map/OsmandRegions.java]: Potentially duplicate WorldRegion from BinaryMapDataObject is not fixed here!

File: OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java (hunk 2 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: fixed

File: OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java (hunk 3 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: fixed

File: OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java (hunk 1 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: fixed

**Pull request merged**